### ISSUE TITLE
GraphObject sub-classes development

### ISSUE THREAD

**Issue by devmsh** [facebook/facebook-php-sdk-v4#401]
The official docs demonstrate some sub-classes like GraphUser, and the code have another unmentioned classes like GraphAlbum, and many other types still haven't classes at all.

I just want to ask about your plan to offer a complete set of GraphObject sub-classes? and if you have any plan to support Edges in addition to the Fields?

I'm really interested to contribute in these points, but I need to hear from you before I add any pull requests?

**Comment by SammyK** [comment]
The documentation for `GraphObject` and the Graph node subtypes can be found [here](https://github.com/facebook/facebook-php-sdk-v4/blob/master/docs/GraphObject.fbmd).

Edges from the Graph API are represented with the [`GraphList` entity](https://github.com/facebook/facebook-php-sdk-v4/blob/master/docs/GraphList.fbmd).

In version 4.2, `GraphObject` will be renamed to `GraphNode` and `GraphList` will be renamed to `GraphEdge` per #320.

I would recommend submitting an issue for each Graph node subtypes you want to create to get a group :+1:. But one off the top of my head the really needs to be created is a `GraphEvent` (but you should still submit an issue to ask the group) :)

One last thing - the PR's should be to the master branch which has the latest version 4.1.

Thanks for being willing to contribute this to the Facebook PHP SDK! :)

**Comment by yguedidi** [comment]
:+1:, and don't forget tests and doc :)

**Comment by unitoneict** [comment]
Sounds great, 

I will start with the GraphEvent and I will create a new PR that contatin both the Sub class and it's tests and doc based on the same standard you have at 4.1

**Comment by SammyK** [comment]
:+1:

**Comment by unitoneict** [comment]
@SammyK 

I found that the Graph API documentation missed many important fields and edges like (declined_count, feed_targeting and others), so I developed a quick code that take the edge metadata=1 response and type map array as an input, and then auto generate both the subclass and the corresponding docs card.

Before I add my PR, I need to be sure about some points:

1. `covor` have no current graphObjectMap like picture, must I create it in the same PR, on a seperate one?
1. `owner` can be mapped to GraphPage, GraphUser or GraphGroup, must I leave it as a GraphObject? or create GraphOwner node that only have getId, and getName function?
1. Why we need to add specific `getGraph{type}` `makeGraph{type}` functions both on `FacebookResponse` and `GraphObjectFactory` rather than handling them using PHP Magic Method?

**Comment by SammyK** [comment]
> I found that the Graph API documentation missed many important fields and edges

Yeah, FB is getting all their Graph API docs setup to be auto-generated so that there aren't any inconsistencies like that. So far they've completed auto-gen for App Node, Page Node, User Node and Video Node & their edges. Once they get the Event Node auto generated, we'll know all the fields & edges & we can add them.

> I developed a quick code that take the edge metadata=1 response and type map array as an input, and then auto generate both the subclass and the corresponding docs card.

I don't know if you're aware of the `getMetaData()` method on the `GraphList` class. The `metadata=1` handling could certainly be better, but just wanted to let you know that the SDK does give you access to it.

> covor have no current graphObjectMap like picture, must I create it in the same PR, on a seperate one?

I think you're safe to put a `GraphCoverPhoto` subtype for the `cover` field in the same PR. :)

> owner can be mapped to GraphPage, GraphUser or GraphGroup, must I leave it as a GraphObject? or create GraphOwner node that only have getId, and getName functions?

Per [the docs](https://developers.facebook.com/docs/graph-api/reference/v2.3/event), the `owner` field returns types `User|Page|Group`. If the docs referenced a type of `Owner`, I would say, "yes, create a GraphOwner", but since the Graph API doesn't have a node called "Owner" I wouldn't create a `GraphOwner`. Instead just do standard `GraphObject`.

> Why we need to add specific getGraph{type} makeGraph{type} functions both on FacebookResponse and GraphObjectFactory rather than handling them using PHP Magic Method?

I suggested something like this back in #159 but this group doesn't like magic. :) Plus, it's better to not use magic when you've got a good IDE you're coding on.

Thanks for your work on this! :)

**Comment by unitoneict** [comment]
I submitted the PR #405, BTW this is my first contribution at Github, and I hope that I did not make any mistake :)

`getMetaData()` is a `GraphList` function that help dealing with pagging and summary, the `metadata=1` is not supported in the SDK and it return a metadata object contain a fields, types, and connections arrays.

for the Owner, I agree with you, I just think to cast the object within the `getOwner` function, but it require additional network connection to determine the type, so I leave it as `GraphObject`.

Once we finished from the Event and I make sure that I did not make any thing wrong, I expect to push additional separated PRs for the rest of the Node after we order their priorities.

**Comment by SammyK** [comment]
> this is my first contribution at Github

Yay! Welcome to the world of open source! :) Warning: it can be addicting!

> I hope that I did not make any mistake :)

One of the best things about open source is that you can make mistakes and people will correct you. @yguedidi corrects my mistakes all the time. We're all human here so you needn't worry about making mistakes - it's expected.

> getMetaData() is a GraphList function that help dealing with pagging and summary, the metadata=1 is not supported in the SDK and it return a metadata object contain a fields, types, and connections arrays.

Ah yes, I forgot I actually had that on my internal TODO list. We def need more metadata support (and specifically debugging support in v2.3+).

**Comment by devmsh** [comment]
@SammyK 

While the `GraphEvent` approved, and I sent the `GraphGroup` PR, can you help me with a to do list for the next 4 or 5 sub classes to be supported?

**Comment by SammyK** [comment]
1. The `GraphUser` node is missing a bunch of fields.
2. We need to audit the existing Graph nodes to make sure they match what is returned when we do a `metadata=1`. The methods should be in the same order too so that it's easier to audit going forward.

Other than that, I say we could do with the following node subtypes pending approval from @yguedidi & @gfosco.

1. `GraphComment`
2. `GraphPhoto`
3. `GraphVideo`
4. `GraphStatus`

**Comment by gfosco** [comment]
:+1:

**Comment by yguedidi** [comment]
:+1: too

**Comment by devmsh** [comment]
Very good.

I liked the way you used `metadata=1` order for auditing, I will update all the old sub classes, reorder the fields, and add any missed fields based on the metadata.

Must I push new PR for each sub type? or I can update all of them at once? If I can update all of them, must I use single commit or one comment for each sub type?

**Comment by SammyK** [comment]
I'd do a separate PR for each subtype so it's easier to review. :)

**Comment by devmsh** [comment]
I refactor the `GraphUser` and it has the same `List` field problem as I comment here https://github.com/facebook/facebook-php-sdk-v4/issues/415#issuecomment-102786175

In addition to that, the `UserGraph` has the following sub types:
userdevice, educationexperience, experience, messagingsuggestedgroup, workexperience, agerange, usercontext, currency, paymentpricepoints, securitysettings, videouploadlimits

I think we need to seperate these partial sub types into separate folder in addition to the current classes like: GraphCoverPhoto, GraphLocation, GraphPicture, and GraphSessionInfo.

By Partial sub types, I mean the ones those did not have direct access from the official refrance menu here: https://developers.facebook.com/docs/graph-api/reference

**Comment by SammyK** [comment]
> I think we need to seperate these partial sub types into separate folder in addition to the current classes like: GraphCoverPhoto, GraphLocation, GraphPicture, and GraphSessionInfo.

Excellent suggestion!

> By partial sub types, I mean the ones those did not have direct access from the official Graph API Reference menu

They could be called many things: "union types", "Facebook primitives", "node attributes". We really need to find out what the people behind the Graph API call these things and then name it that. Lol.

The thing is I don't have any direct contact with any Graph API people. @gfosco Do you think there is anyone from the Graph API team who would be willing to answer questions like this for us going forward to make sure the PHP SDK keeps a solid [DSL](http://en.wikipedia.org/wiki/Domain-specific_language) with Facebook? :)

**Comment by devmsh** [comment]
I review the the official documentation, and this is the only thing I found about that

> The main difference between a root node and a non-root node is that root nodes can be queried directly, while non-root nodes can be queried via root nodes or edges.

Waiting from @gfosco

**Comment by SammyK** [comment]
"Non-root node," it's got a nice ring to it. Said no one. Ever. Lol.

Nice find in the docs! Maybe the Graph API kids call it something else internally.

**Comment by browner12** [comment]
not sure if this is the correct issue to discuss this in, so redirect me if need be.

what do people think about taking a page from the Laravel Eloquent playbook, and using the `__get()` method to access primitive variables? this would mean instead of things like `$user->getId()` and `$user->getFirstName()`, we could use `$user->id` and `$user->firstName`. this would clean up the entity classes quite a bit, be a little more fluent, and prevent a lot of extra coding if new fields are added in the future.

the PR would be relatively simple, too, i believe. add `__get()` to the `Collection` class and call the `getProperty()` method from it, and then remove all of the specific getters from the GraphNodes classes.

thoughts?

**Comment by devmsh** [comment]
Magic again :D 

I think that we must rethink of that request!, If the IDE support is the concern, what about using @method annotation? or use __get() as a fallback for the not directly supported fields?

**Comment by SammyK** [comment]
Thanks for the feedback @browner12! The magic-access of properties has been discussed a lot. I'm still :-1: for it but if enough people want it, I think it should be added. :)

**Comment by yguedidi** [comment]
I'm still :-1: because:
- [as I already show you](https://github.com/facebook/facebook-php-sdk-v4/issues/429#issuecomment-101260649), magic is a pain
- I don't think it's makes sense to mix non-magic and magic
- About `@method` annotation, if one is missed, it will require a PR, like now
- Inspired from above point, magic will stop some contributions :)

**Comment by SammyK** [comment]
> magic will stop some contributions

That's probably the strongest argument against magic yet. The node subtype contributions have been the most common entry point for new contributors in the PHP SDK.

And I'm with you on the other points as well. :)

**Comment by browner12** [comment]
looks like i reopened a can of worms. oops. well, here's my 2 cents on it.

to start, i was in the no magic methods camp for a long time, but after writing sooo much boilerplate code I've come around to it.

I think we can all agree magic code is bad. magic means something happens and we don't understand how it happens, or can't easily figure out how it happens with a little bit of digging.  I think it's really bad luck these methods were named magic methods, because in reality we do understand how they work, and a developer can easily check the source to see what happens when they are called. Technically `__construct()` is a magic method but I don't think anyone would even begin to argue against the use of a constructor.

If we truly don't want any *magic* going on behind the scenes, why don't we just make each field a `public` property on the object? We don't want that because it gives us no control over the properties.  So now we mark all the field properties as `private` and each of them gets its own getter. Most of the getters just return the straight value but some of them do some processing (some might say magic) and automatically convert a birthday field to a `Datetime` or `Carbon` object. We also want to convert other datetimes like `created_time` and `updated_time` to `Datetime` objects, so we copy the code to create `Datetime`s to each of these getters as well.  We don't want that because it means we're duplicating a ton of code. So we create a super method called `getProperty()` and store all of our fields in a private property array called `items`.  I would say all of these progressions are good steps forward and add a little bit of logic/processing/magic (whatever you want to call it) but do so in a transparent way. Now each of the getters calls `getProperty()`, but we don't want that because again we're duplicating code. So I would argue that adding

```php
public function __get($key)
{
    return $this->getProperty($key);
}
```

is simply the next step. It reduces a *TON* of boilerplate code, and is simple to understand. 

in fact, I would argue that there's more *magic* going on in the `getProperty()` method than there is in the `__get()` method.

TLDR it's only magic if you don't understand how it works.

---

> as I already show you, magic is a pain

if you're referencing the fact here that PHP versions below 5.1 can't handle it, I really don't see how that is a point. The package specifically states that it requires PHP >=5.4

> I don't think it's makes sense to mix non-magic and magic

I'm not sure I truly understand your point here. do you have a reason behind this? and like I said before `__construct()` and `__toString()` are magic, and they're mixed in there.

> magic will stop some contributions

as much as I kind of hate myself for saying this, is that a bad thing? are we really doing a service by leaving something in only so contributors can add boilerplate code? and I can also say that Laravel's Eloquent ORM `Model` class uses `__get()` and still has plenty of contribution.

---

BTW, does this repo have an IRC or slack channel?  might be a better place for this discussion

**Comment by yguedidi** [comment]
@browner12 Thanks for your 2 cents, now mine:

> I would argue that there's more magic going on in the getProperty() method

You're totaly right, and thats why I'll send a proposal to refactor that once @gfosco gives us his thoughts about BC for 5.0 (ref. #436)

> if you're referencing the fact here that PHP versions below 5.1 can't handle it, I really don't see how that is a point. The package specifically states that it requires PHP >=5.4

I'm refering to the second `var_dump()` call, which says that `empty($obj->foo)` returns true...

> I'm not sure I truly understand your point here. do you have a reason behind this? and like I said before `__construct()` and `__toString()` are magic, and they're mixed in there.

I know `__construct()` and `__toString()` are also magic method.
When I talk about magic and the fact that I don't like that, I mean magic methods which open the class API.
Class API should be explict and restrict to only defined use cases.

> as much as I kind of hate myself for saying this, is that a bad thing?

Yes it is, there is no perfect project :)

> are we really doing a service by leaving something in only so contributors can add boilerplate code?

All contributions are welcome, we are humans, we can forgot many things. And if because of our lyziness as developers we add magic but also add all above problems, I'm :-1: 

>  Laravel's Eloquent ORM

I'm personaly not a big fan of Laravel. I think, and it's only my opinio, that Laravel use all magic/trick/shortchuts available in PHP because it's a framework dedicated to RAD, and I have no problem with that.
But we are not a framework, but a SDK, a library that should be explicit about what it provides to its users, developers.

> does this repo have an IRC or slack channel?

:+1: for a slack channel, or even better, a [gitter](https://gitter.im/) room

**Comment by SammyK** [comment]
> Laravel's Eloquent ORM

I love Laravel and use it for most of my projects. But the Eloquent ORM is for maintaining mutable objects for persistent storage and has very different behavior from the immutable Graph node response objects we're talking about here.

But I'm not making too many cases in the argument b/c I think this falls into a "tabs vs spaces" religious camp. :)

```php
	// <- a tab
    // <- 4 spaces
```

```php
echo $node->['id']; // <- array access (existing magic)
echo $node->id; // <- __get() ("artisan" magic) ;)
```

**Comment by browner12** [comment]
thanks for the responses guys. some interesting points to think about.

**Comment by devmsh** [comment]
I just want to ask why we documented all the `GraphNode` sub classes in the same file? while I add more and more sub classes, I think we must move the docs for each sub class in a separated file?

This will make it more maintainable?

**Comment by SammyK** [comment]
I'm open to having them in separate files, but for now I think we should keep it the same so that @gfosco won't have any trouble loading the docs and we can get 5.0 out the door. :)

**Comment by devmsh** [comment]
mmmm I can not wait to finalize the publishing of the new version, naming issue and #437 to go on!

It take more time than expected :(

**Comment by gfosco** [comment]
@yguedidi I'm :+1: on breaking things, :-1: on Magic. 

Sorry that the 5.0 is release is taking longer than expected.  I have soo much on my plate, and no one else to help load the docs.  We can technically tag 5.0 before the docs get loaded, though, that is an option I won't dismiss..

**Comment by yguedidi** [comment]
@gfosco Thanks!

I think we shouldn't tag it, even without the doc loaded, see #447

**Comment by SammyK** [comment]
> We can technically tag 5.0 before the docs get loaded, though, that is an option I won't dismiss..

Related to my comments on #447, I think tagging & branching a stable 5.0 release would be a great idea and we can star BC on the master branch which will eventually become v6.0.

**Comment by gfosco** [comment]
Closing this.. if there are still points to discuss, please start a new issue.
